### PR TITLE
Analysis speed improvements - call graph, topological order of blocks etc.

### PR DESCRIPTION
**Pull request by roberthusak**
Introducing call graph, priority queue and heuristics to analysis blocks ordering resulting in faster analysis.

### CODE REVIEW

File: src/Peachpie.CodeAnalysis/Utilities/PriorityQueue.cs [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+        /// <summary>
+        /// Adds an object to the priority queue.
+        /// </summary>
+        internal void Push(T value)
+        {
+            // Increase the size of the array if necessary.
+            if (_count == _heap.Length)
+            {
+                T[] temp = new T[_count * 2];
```
Comment (jakubmisek) [inline src/Peachpie.CodeAnalysis/Utilities/PriorityQueue.cs]: line 65 - 70: `Array.Resize`, eventualy at least use `Array.Copy`

File: src/Peachpie.CodeAnalysis/Utilities/PriorityQueue.cs [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+            {
+                Debug.Assert(_count > 0);
+                return _heap[0];
+            }
+        }
+
+        /// <summary>
+        /// Adds an object to the priority queue.
+        /// </summary>
+        internal void Push(T value)
```
Comment (jakubmisek) [inline src/Peachpie.CodeAnalysis/Utilities/PriorityQueue.cs]: is it thread safe? always called within lock ?

File: src/Peachpie.CodeAnalysis/Utilities/PriorityQueue.cs [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+            }
+
+            _heap[index] = value;
+            _count++;
+        }
+
+        /// <summary>
+        /// Removes the first node (i.e., the logical root) from the heap.
+        /// </summary>
+        internal void Pop()
```
Comment (jakubmisek) [inline src/Peachpie.CodeAnalysis/Utilities/PriorityQueue.cs]: thread safety alright?

**Comment by bfistein** [comment]
Great stuff, thanks a lot @roberthusak. Gotta sign that CLA though ;)

File: src/Peachpie.CodeAnalysis/Utilities/PriorityQueue.cs [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+        {
+            Debug.Assert(_count != 0);
```
Comment (jakubmisek) [inline src/Peachpie.CodeAnalysis/Utilities/PriorityQueue.cs]: `throw` if _count <= 0 instead, to be sure in release

File: src/Peachpie.CodeAnalysis/Utilities/PriorityQueue.cs [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+        /// <summary>
+        /// Gets the first or topmost object in the priority queue, which is the
+        /// object with the minimum value.
+        /// </summary>
+        public T Top
+        {
+            get
+            {
+                Debug.Assert(_count > 0);
```
Comment (jakubmisek) [inline src/Peachpie.CodeAnalysis/Utilities/PriorityQueue.cs]: `throw` instead

File: src/Peachpie.CodeAnalysis/Utilities/PriorityQueue.cs [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        private T[] _heap;
+        private int _count;
+        private IComparer<T> _comparer;
+        private const int DefaultCapacity = 6;
+
+        public PriorityQueue(IComparer<T> comparer)
+        {
+            _heap = new T[DefaultCapacity];
+            _count = 0;
+            _comparer = comparer;
```
Comment (jakubmisek) [inline src/Peachpie.CodeAnalysis/Utilities/PriorityQueue.cs]: `?? throw new ArgumentNullException`

File: src/Peachpie.CodeAnalysis/Utilities/PriorityQueue.cs [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+    /// order is not guaranteed to be stable.
+    /// 
+    /// The original implementation was taken from the Microsoft .NET
+    /// Framework Reference Source, author: Niklas Borson (niklasb).
+    /// </remarks>
+    internal class PriorityQueue<T>
+    {
+        private T[] _heap;
+        private int _count;
+        private IComparer<T> _comparer;
```
Comment (jakubmisek) [inline src/Peachpie.CodeAnalysis/Utilities/PriorityQueue.cs]: `readonly`

File: src/Peachpie.CodeAnalysis/Utilities/PriorityQueue.cs [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+    /// <remarks>
+    /// Push and Pop are each O(log N). Pushing N objects and them popping
+    /// them all is equivalent to performing a heap sort and is O(N log N).
+    /// Multiple different items of the same priority are allowed, but their
+    /// order is not guaranteed to be stable.
+    /// 
+    /// The original implementation was taken from the Microsoft .NET
+    /// Framework Reference Source, author: Niklas Borson (niklasb).
+    /// </remarks>
+    internal class PriorityQueue<T>
```
Comment (jakubmisek) [inline src/Peachpie.CodeAnalysis/Utilities/PriorityQueue.cs]: found issue in Microsoft impl - https://stackoverflow.com/questions/44221454/bug-in-microsofts-internal-priorityqueuet

Do we fix this?

File: src/Peachpie.CodeAnalysis/Utilities/DistinctQueue.cs [hunk 9]
```diff
@@ -5,18 +5,27 @@
...
+using Peachpie.CodeAnalysis.Utilities;
 
 namespace Pchp.CodeAnalysis.Utilities
 {
     /// <summary>
-    /// Represents queue where items are enqueued just once and queue can be accessed in parralel.
+    /// Represents priority queue where items are enqueued just once and queue can be accessed in parallel.
     /// </summary>
-    internal sealed class DistinctQueue<T>
+    internal sealed class DistinctQueue<T> where T : BoundBlock
```
Comment (jakubmisek) [inline src/Peachpie.CodeAnalysis/Utilities/DistinctQueue.cs]: I would remove the condition (it makes `DistinctQueue` unusable for anything else).

File: src/Peachpie.CodeAnalysis/Utilities/DistinctQueue.cs [hunk 10]
```diff
@@ -5,18 +5,27 @@
...
     {
         readonly object _syncRoot = new object();
 
-        readonly HashSet<T> _set = new HashSet<T>();
```
Comment (jakubmisek) [inline src/Peachpie.CodeAnalysis/Utilities/DistinctQueue.cs]: `IsEnqueued` does not seems right - now we cannot use the single instance of T within more than one `DistinctQueue` and `DistinctQueue` would require its items to always have `IsEnqueued` property. Also the class `DistinctQueue` is now unusable for general use.

File: src/Peachpie.CodeAnalysis/FlowAnalysis/AnalysisVisitor.cs [hunk 11]
```diff
@@ -89,9 +89,14 @@ private void TraverseToBlock(object edgeLabel, FlowState/*!*/state, BoundBlock/*
                 // merge and check whether state changed
                 state = state.Merge(targetState);   // merge states into new one
 
-                if (state.Equals(targetState))
+                if (state.Equals(targetState) && !(target is ExitBlock))
```
Comment (jakubmisek) [inline src/Peachpie.CodeAnalysis/FlowAnalysis/AnalysisVisitor.cs]: Special type condition for `ExitBlock` is just not nice.

File: src/Peachpie.CodeAnalysis/FlowAnalysis/AnalysisVisitor.cs [hunk 12]
```diff
@@ -327,23 +332,32 @@ public override void VisitCFGExitBlock(ExitBlock x)
...
+                if (rtype != exit._lastReturnTypeMask || wasNotAnalysed)
                 {
                     exit._lastReturnTypeMask = rtype;
                     var subscribers = exit.Subscribers;
                     if (subscribers.Count != 0)
                     {
                         lock (subscribers)
                         {
-                            subscribers.ForEach(_worklist.Enqueue);
+                            subscribers.ForEach(subscriber => _worklist.PingReturnUpdate(exit, subscriber));
```
Comment (jakubmisek) [inline src/Peachpie.CodeAnalysis/FlowAnalysis/AnalysisVisitor.cs]: this creates an ugly closure - change it to simple `foreach (var s in subscribers) _worlist.Ping...;` pls

File: src/Peachpie.CodeAnalysis/Utilities/PriorityQueue.cs (hunk 2 above)
Comment (roberthusak) [reply src/Peachpie.CodeAnalysis/Utilities/PriorityQueue.cs]: No, ```PriorityQueue``` is a simple heap with no locking. Thread safety is provided by ```DistinctQueue```.

File: src/Peachpie.CodeAnalysis/Utilities/PriorityQueue.cs (hunk 3 above)
Comment (roberthusak) [reply src/Peachpie.CodeAnalysis/Utilities/PriorityQueue.cs]: ditto

File: src/Peachpie.CodeAnalysis/Utilities/PriorityQueue.cs (hunk 8 above)
Comment (roberthusak) [reply src/Peachpie.CodeAnalysis/Utilities/PriorityQueue.cs]: Wow.. I should've checked that. Anyway, it's fixed now in b34ed05.

File: src/Peachpie.CodeAnalysis/Utilities/DistinctQueue.cs (hunk 10 above)
Comment (roberthusak) [reply src/Peachpie.CodeAnalysis/Utilities/DistinctQueue.cs]: OK, I've probably over-optimised this bit O:-) It's reverted back to the usage of ```HashSet```.

File: src/Peachpie.CodeAnalysis/FlowAnalysis/AnalysisVisitor.cs (hunk 11 above)
Comment (roberthusak) [reply src/Peachpie.CodeAnalysis/FlowAnalysis/AnalysisVisitor.cs]: Replaced by ```BoundBlock.ForceRepeatedAnalysis``` in 529d61b.

**Comment by roberthusak** [comment]
@jakubmisek Thanks for the comments. I've reflected them in the recent commits.

File: src/Peachpie.CodeAnalysis/Utilities/DistinctQueue.cs [hunk 13]
```diff
@@ -44,10 +50,10 @@ public bool Enqueue(T value)
 
             lock (_syncRoot)
             {
-                if (!value.IsEnqueued)
+                if (!_set.Contains(value))
```
Comment (jakubmisek) [inline src/Peachpie.CodeAnalysis/Utilities/DistinctQueue.cs]: `_set.Contains(x) and _set.Add(x)` can be simplified to single lookup `!_set.Add(x)`

Comment (roberthusak) [reply src/Peachpie.CodeAnalysis/Utilities/DistinctQueue.cs]: OK, fixed in 7d635e6.

**Pull request merged**